walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,127 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {   
+    // Make optional attendees mandatory if there are no mandatory attendees.
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    if (attendees.size() == 0 && optionalAttendees.size() > 0) {
+      attendees = optionalAttendees;
+      optionalAttendees = new ArrayList<String>();
+    }
+
+    if (optionalAttendees.size() == 0) {
+      return FindMeetingTimes(events, request, attendees);
+    }
+    Collection<String> allAttendees = merge(attendees, optionalAttendees);
+    Collection<TimeRange> meetingTimes = FindMeetingTimes(events, request, allAttendees);
+    if (meetingTimes.size() == 0) {
+      return FindMeetingTimes(events, request, attendees);
+    }
+    return meetingTimes;
+  }
+
+  /* Finds Meeting Times for All People Given */
+  private Collection<TimeRange> FindMeetingTimes(Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Create list of all events an attendee for this meeting is going to.  
+    ArrayList<TimeRange> eventsAttendedTimes = new ArrayList();
+    for (Event event : events) {
+      if (sharesAttendee(event, attendees)) {
+        eventsAttendedTimes.add(event.getWhen());
+      }
+    }
+    // Sort eventsAtendedTimes in ascending order of event start time.
+    Collections.sort(eventsAttendedTimes, TimeRange.ORDER_BY_START);   
+    // Merge overlapping TimeRange's in eventAttededTimes.
+    Stack<TimeRange> stack = new Stack<TimeRange>();
+    if (eventsAttendedTimes.size() > 0) {
+      stack.push(eventsAttendedTimes.get(0)); 
+    }
+    TimeRange nextTime;   
+    for (int i = 1; i < eventsAttendedTimes.size(); i++) {
+      nextTime = eventsAttendedTimes.get(i);
+      if (stack.peek().overlaps(nextTime)) {
+        if (nextTime.end() > stack.peek().end()) {
+          // pop from stack, use pop's start
+          int start = stack.pop().start();
+          stack.push(TimeRange.fromStartEnd(start, nextTime.end(), false));
+        }
+      }
+      else {
+        stack.push(nextTime);
+      }
```

Comment:
[Optional] This logic for merging overlapping time ranges is complex enough that it might warrant it's own function.  It would also help avoid such deep nesting.